src/gui/static/src/app/services/hw-wallet.service.ts

```diff
@@ -0,0 +1,350 @@
+      });
+    });
+  }
+
+  backup(): Observable<OperationResult> {
+    return this.cancelLastAction().flatMap(() => {
+      const requestId = this.createRandomIdAndPrepare();
+      window['ipcRenderer'].send('hwBackupDevice', requestId);
+
+      return new Observable(observer => {
+        this.eventsObservers.set(requestId, observer);
+      });
+    });
+  }
+
+  wipe(): Observable<OperationResult> {
+    return this.cancelLastAction().flatMap(() => {
+      const requestId = this.createRandomIdAndPrepare();
+      window['ipcRenderer'].send('hwWipe', requestId);
+
+      return new Observable(observer => {
+        this.eventsObservers.set(requestId, observer);
+      });
+    });
+  }
+
+  signMessage(addressIndex: number, message: string, currentSignature: number, totalSignatures: number): Observable<OperationResult> {
+    return this.cancelLastAction().flatMap(() => {
+      const requestId = this.createRandomIdAndPrepare();
+      this.signingTx = true;
+      this.currentSignature = currentSignature;
+      this.totalSignatures = totalSignatures;
+      window['ipcRenderer'].send('hwSignMessage', requestId, addressIndex, message);
+
+      return new Observable(observer => {
+        this.eventsObservers.set(requestId, observer);
+      });
+    });
+  }
+
+  checkIfCorrectHwConnected(firstAddress: string): Observable<boolean> {
+    return this.getAddresses(1, 0).flatMap(
+      response => {
+        if (response.rawResponse[0] !== firstAddress) {
+          return Observable.throw({
+            result: OperationResults.IncorrectHardwareWallet,
+            rawResponse: '',
+          });
+        }
+
+        return Observable.of(true);
+      },
+    );
+  }
+
+  private getAddressesRecursively(index: number, addresses: string[]): Observable<string[]> {
+    let chain: Observable<any>;
+    if (index > 0) {
+      chain = this.getAddressesRecursively(index - 1, addresses).first();
+    } else {
+      chain = Observable.of(1);
+    }
+
+    chain = chain.flatMap(() => {
+      return this.getAddresses(1, index)
+      .map(response => {
+        addresses.push(response.rawResponse[0]);
+
+        return addresses;
+      });
+    });
+
+    return chain;
+  }
+
+  private createRandomIdAndPrepare() {
+    this.changingPin = false;
+    this.signingTx = false;
+
+    return Math.floor(Math.random() * 4000000000);
+  }
+
+  private dispatchEvent(requestId: number, rawResponse: any, success: boolean) {
+    if (this.eventsObservers.has(requestId)) {
+      if (!rawResponse.error && success) {
+        this.eventsObservers.get(requestId).next({
+          result: OperationResults.Success,
+          rawResponse: rawResponse,
+        });
+      } else {
+        const responseContent = rawResponse.error ? rawResponse.error : rawResponse;
+        let result: OperationResults;
+
+        if (typeof responseContent === 'string' && (responseContent as string).includes('failed or refused')) {
```

Comment:
could be simplified with a map of string -> operationresult

Reply:
I'm not sure if the texts should be converted to lowercase in that way, as I do not know if at some point some messages returned by the hardware wallet will be showed directly in the UI. I also did not use a map because the strings I am using are not the exact responses returned by the hardware wallet (I would have to check the code of the js library to be sure again about the responses and in fact I would prefer to change that part on the JS library). However, I did simplify the code a bit